### PR TITLE
feat: add basic proxy

### PR DESCRIPTION
**Pull request by igorbernstein2**
Stacked PRs:
 * #403
 * #402
 * #401
 * __->__#400
 * #399
 * #398

--- --- ---

### CODE REVIEW

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/commands/Serve.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    if (adminChannel == null) {
+      adminChannel =
+          ManagedChannelBuilder.forAddress(adminEndpoint.getName(), adminEndpoint.getPort())
+              .userAgent(userAgent)
+              .disableRetry()
+              .build();
+    }
+
+    Map<String, ServerCallHandler<byte[], byte[]>> serviceMap =
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/commands/Serve.java:93]: Maybe add a comment specifying where to find valid service names

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+package com.google.cloud.bigtable.examples.proxy.core;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+
+/** A factory to pairing of an incoming server call to an outgoing client client. */
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java:25]: - `s/A factory to pairing/A factory pairing` or `s/A factory to pairing/A factory to pair`
- `s/outgoing client client/outgoing client call`

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ */
+package com.google.cloud.bigtable.examples.proxy.core;
+
+import io.grpc.ClientCall;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.Status;
+import javax.annotation.concurrent.GuardedBy;
+
+/** A per RPC proxy. */
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java:24]: NIT: should we specify it is specifically grpc? (and I guess as a side question: is it actually specifically grpc?)

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+class CallProxy<ReqT, RespT> {
+  final RequestProxy serverCallListener;
+  final ResponseProxy clientCallListener;
+
+  public CallProxy(ServerCall<ReqT, RespT> serverCall, ClientCall<ReqT, RespT> clientCall) {
+    serverCallListener = new RequestProxy(clientCall);
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java:30]: NIT: add a short comment specifying what is `clientCall`, `clientCallListener`, `serverCall` and `serverCallListener`

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  @Override
+  public ServerCall.Listener<ReqT> startCall(ServerCall<ReqT, RespT> serverCall, Metadata headers) {
+    CallOptions callOptions = CallOptions.DEFAULT;
+
+    ClientCall<ReqT, RespT> clientCall =
+        channel.newCall(serverCall.getMethodDescriptor(), callOptions);
+
+    CallProxy<ReqT, RespT> proxy = new CallProxy<>(serverCall, clientCall);
+    clientCall.start(proxy.clientCallListener, headers);
+    serverCall.request(1);
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java:42]: Add a comment for why we call `.request(1)` here

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    clientCallListener = new ResponseProxy(serverCall);
+  }
+
+  private class RequestProxy extends ServerCall.Listener<ReqT> {
```
Comment (andre-sampaio) [inline java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java:34]: Maybe add a brief comment explaining the flow here (or perhaps to `onMessage`). Same for Response proxy.

IIUC onMessage is called multiple times until a full request message is received (I think this is what turns `clientCall.isReady() == true`?), and then it sends that to the server. Is this right?

**Review by andre-sampaio — APPROVED** [review]

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java (hunk 5 above)
Comment (igorbernstein2) [reply java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/ProxyHandler.java:42]: done

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java (hunk 4 above)
Comment (igorbernstein2) [reply java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java:30]: done

File: java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java (hunk 6 above)
Comment (igorbernstein2) [reply java/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java:34]: done

**Comment by igorbernstein2** [comment]
closing in favor https://github.com/GoogleCloudPlatform/java-docs-samples/pull/9689

**Pull request closed without merging**